File: packages/java/src/main/java/dev/kreuzberg/PageContent.java
package dev.kreuzberg;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

/**
 * Content for a single page/slide.
 *
 * <p>
 * When page extraction is enabled, documents are split into per-page content
 * with associated tables and images mapped to each page.
 *
 * @param pageNumber
 *            the page number (1-indexed)
 * @param content
 *            the text content for this page
 * @param tables
 *            tables found on this page
 * @param images
 *            images found on this page
 * @param hierarchy
 *            hierarchy information for the page, or null
 * @param isBlank
 *            whether this page is blank, or null
 * @param layoutRegions
 *            layout regions detected on this page, or null
 */
public record PageContent(@JsonProperty("page_number") int pageNumber, @JsonProperty("content") String content,
		@JsonDeserialize(contentAs = Table.class) @JsonProperty("tables") List<Table> tables,
		@JsonDeserialize(contentAs = ExtractedImage.class) @JsonProperty("images") List<ExtractedImage> images,
		@JsonProperty("hierarchy") PageHierarchy hierarchy, @JsonProperty("is_blank") Boolean isBlank,
		@JsonDeserialize(contentAs = LayoutRegion.class) @JsonProperty("layout_regions") List<LayoutRegion> layoutRegions) {
	@JsonCreator
	public PageContent(@JsonProperty("page_number") int pageNumber, @JsonProperty("content") String content,
			@JsonProperty("tables") List<Table> tables, @JsonProperty("images") List<ExtractedImage> images,
			@JsonProperty("hierarchy") PageHierarchy hierarchy, @JsonProperty("is_blank") Boolean isBlank,
			@JsonDeserialize(contentAs = LayoutRegion.class) @JsonProperty("layout_regions") List<LayoutRegion> layoutRegions) {
		this.pageNumber = pageNumber;
		this.content = content != null ? content : "";
		this.tables = tables != null ? Collections.unmodifiableList(tables) : List.of();
		this.images = images != null ? Collections.unmodifiableList(images) : List.of();
		this.hierarchy = hierarchy;
		this.isBlank = isBlank;
		this.layoutRegions = layoutRegions != null ? Collections.unmodifiableList(layoutRegions) : List.of();
	}

	/**
	 * Get the hierarchy information for this page.
	 *
	 * @return hierarchy, or empty if not available
	 */
	public Optional<PageHierarchy> getHierarchy() {
		return Optional.ofNullable(hierarchy);
	}

	/**
	 * Get whether this page is blank (contains no meaningful content).
	 *
	 * @return true if blank, false otherwise, empty if not applicable
	 */
	public Optional<Boolean> getIsBlank() {
		return Optional.ofNullable(isBlank);
	}

	/**
	 * Get the layout regions detected on this page.
	 *
	 * @return layout regions, or empty if layout detection was not enabled
	 */
	public Optional<List<LayoutRegion>> getLayoutRegions() {
		return layoutRegions.isEmpty() ? Optional.empty() : Optional.of(layoutRegions);
	}
}

File: packages/php/src/Types/PageContent.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Content for a single page/slide.
 *
 * When page extraction is enabled, documents are split into per-page content
 * with associated tables and images mapped to each page.
 *
 * @property-read int $pageNumber Page number (1-based)
 * @property-read string $content Page text content
 * @property-read array<Table> $tables Tables found on this page
 * @property-read array<ExtractedImage> $images Images found on this page
 * @property-read ?PageHierarchy $hierarchy Hierarchy information for the page
 * @property-read ?bool $isBlank Whether this page is blank
 * @property-read array<LayoutRegion> $layoutRegions Layout regions detected on this page
 */
readonly class PageContent
{
    /**
     * @param array<Table> $tables
     * @param array<ExtractedImage> $images
     * @param array<LayoutRegion> $layoutRegions
     */
    public function __construct(
        public int $pageNumber,
        public string $content,
        public array $tables = [],
        public array $images = [],
        public ?PageHierarchy $hierarchy = null,
        public ?bool $isBlank = null,
        public array $layoutRegions = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var int $pageNumber */
        $pageNumber = $data['page_number'] ?? 0;

        /** @var string $content */
        $content = $data['content'] ?? '';

        /** @var array<array<string, mixed>> $tablesData */
        $tablesData = $data['tables'] ?? [];

        /** @var array<array<string, mixed>> $imagesData */
        $imagesData = $data['images'] ?? [];

        /** @var ?array<string, mixed> $hierarchyData */
        $hierarchyData = isset($data['hierarchy']) && is_array($data['hierarchy']) ? $data['hierarchy'] : null;

        /** @var ?bool $isBlank */
        $isBlank = isset($data['is_blank']) && is_bool($data['is_blank']) ? $data['is_blank'] : null;

        /** @var array<array<string, mixed>> $layoutRegionsData */
        $layoutRegionsData = isset($data['layout_regions']) && is_array($data['layout_regions'])
            ? $data['layout_regions']
            : [];

        return new self(
            pageNumber: $pageNumber,
            content: $content,
            tables: array_map(
                /** @param array<string, mixed> $table */
                static fn (array $table): Table => Table::fromArray($table),
                $tablesData,
            ),
            images: array_map(
                /** @param array<string, mixed> $image */
                static fn (array $image): ExtractedImage => ExtractedImage::fromArray($image),
                $imagesData,
            ),
            hierarchy: $hierarchyData !== null ? PageHierarchy::fromArray($hierarchyData) : null,
            isBlank: $isBlank,
            layoutRegions: array_map(
                /** @param array<string, mixed> $region */
                static fn (array $region): LayoutRegion => LayoutRegion::fromArray($region),
                $layoutRegionsData,
            ),
        );
    }
}
